doc/reference/trd104-syscalls.md

```diff
@@ -510,13 +510,13 @@ kernel returns `Failure` with an error code of `NODEVICE`.
 
 The Read-Write Allow system call class is how a userspace process
 shares buffer with the kernel that the kernel can read and write. When
-userspace shares a buffer, it can no longer access it. Calling a
+userspace shares a buffer, it must no longer modify it. Calling a
```

Comment:
I think we want to say access. The kernel can write to the buffer. We don't want userspace to view intermediate writes. So I think "it must no longer access it"?

Reply:
It seems possible there are cases where userspace is willing to accept the possibility of intermediate writes in exchange for a reduction in overhead.

As written, apps are required to unallow and reallow buffers every time they want to read from them. But for many capsules, this is unnecessary - for example, an ADC capsule can promise not to write to an allowed buffer after issuing a sampling_complete() callback. This would allow an application reading results out of that allowed buffer to save the overhead of two system calls to unallow and re-allow the buffer.

Of course, in the general case this is not good practice, and application authors should be careful to investigate their particular use case before trying this. But for a given capsule it is possible to make it completely safe for userspace to do this, and I don't see why we should forbid it in documentation. As the documentation is written, it seems like we would be compelled to request changes on any userspace library submitted to `libtock-c` that relies on this.

Reply:
I agree that allow-based swapping is a significant overhead, and we should avoid when it's safe to do so. But I think the problem is that there's a coupling between the liveness of the slice in the kernel and particular calling semantics of commands. To take your ADC example, how do we know that the ADC capsule "promises" that? It can't be that a *capsule* does -- that's an implementation. It has to be that it's specified in the device syscall API that an implementer of the ADC system calls won't do so. 

But historically, our ability or inclination to precisely specify system call APIs has been not great. It's pretty much "this is what the capsule implements, here's the libtock wrapper for it". Since we've been unable to precisely specify the liveness of buffers or safety of operations in system call APIs, we have two choices:
  1. Specify the liveness/ownership of buffers at the system call class level (this is what TRD 104 does).
  2. Change course and precisely specify the system call APIs of different device drivers and then check that our particular implementations follow those specifications.

I took the approach of 1. in 104 because I thought it was wishful thinking to achieve 2. But if we can do 2, then I think it would be totally reasonable for 104 to lay out the criteria for when allowed slices can be accessed by userspace, as well as how a system call API should specify this.

Reply:
> To take your ADC example, how do we know that the ADC capsule "promises" that? It can't be that a capsule does -- that's an implementation. It has to be that it's specified in the device syscall API that an implementer of the ADC system calls won't do so.

I think it is fine for a device syscall API to specify something like that if needed. I also think that in many real-world use cases of Tock the author of the capsule will be the same as the author of the userspace library that interacts with the capsule, and those users are happy to rely on the implementation as the "promise". Even if we cannot rely on that promise for capsules/userspace code submitted  upstream, we should not have documentation that suggests those users should not make optimizations which are in reality perfectly safe to make if done correctly.

> But historically, our ability or inclination to precisely specify system call APIs has been not great. It's pretty much "this is what the capsule implements, here's the libtock wrapper for it". Since we've been unable to precisely specify the liveness of buffers or safety of operations in system call APIs, we have two choices:
>
> 1. Specify the liveness/ownership of buffers at the system call class level (this is what TRD 104 does).
> 2. Change course and precisely specify the system call APIs of different device drivers and then check that our particular implementations follow those specifications.

I would suggest something like:

3. Specify that device drivers lose access to allowed buffers unless the system call API of that particular device driver promises safe access to allowed buffers under certain conditions.

This prevents us from closing the door on these optimizations in the future, without forcing us to undertake precisely specifying every system call API (which I agree is probably wishful thinking)